src/shared/reboot-util.c

```diff
@@ -107,3 +108,43 @@ int shall_restore_state(void) {
 
         return r > 0 ? ret : true;
 }
+
+static int xen_kexec_loaded(void) {
+#if HAVE_XENCTRL
+        xc_interface *xch;
+        int r;
+
+        r = dlopen_xenctrl();
+        if (r < 0)
+                return r;
+
+        if (access("/proc/xen", F_OK) < 0)
+                return log_debug_errno(errno, "Cannot access /proc/xen: %m");
```

Comment:
this check is much more light-weight than the dlopen_xenctrl() check, hence please do this first.